Fix scheduler thread wakeups

veerajsk opened the pull request:
It was possible for the scheduler thread to wait on its condition
variable while the condition variable was notified during both
termination & addition of a job to the job queue due to race conditions
resulting from the unlocked manipulation of the running flag & the job
queue. Fix this by grabbing the wait mutex before modifying said
variables.

veerajsk committed:
Fix scheduler thread wakeups

It was possible for the scheduler thread to wait on its condition
variable while the condition variable was notified during both
termination & addition of a job to the job queue due to race conditions
resulting from the unlocked manipulation of the running flag & the job
queue. Fix this by grabbing the wait mutex before modifying said
variables.

xffox commented on alica_engine/include/engine/scheduler/Scheduler.h:
```diff
@@ -66,9 +69,12 @@ class Scheduler
+            std::unique_lock<std::mutex> lock(_mutex);
+            _running = false;
+        }
         _cv.notify_one();
```
Hm, i think you might still loose the `_cv` notification on around line 100 before the lock. Maybe then you need to check the values before the `wait()`, but under the lock.

veerajsk replied:
that's okay, since `running` will be false in that case & we will never wait on the condition variable

veerajsk replied:
the predicate passed to the wait() function is called before going into wait but under the lock

xffox replied:
right

xffox commented:
The job queue is not under the lock, are there performance concerns?

veerajsk commented:
> The job queue is not under the lock, are there performance concerns?

The queue itself is multi thread safe i.e. it has its own internal lock that it uses to guard access

xffox commented:
> > The job queue is not under the lock, are there performance concerns?
> 
> The queue itself is multi thread safe i.e. it has its own internal lock that it uses to guard access

Right, i mean with one more lock it's slower, would it affect someone?

abhi1287 commented on alica_engine/include/engine/scheduler/Scheduler.h:
```diff
@@ -53,7 +53,10 @@ class Scheduler
-        _jobQueue.emplace(jobId, std::move(jobCb), std::move(repeatInterval));
+        {
+            std::unique_lock<std::mutex> lock(_mutex);
+            _jobQueue.emplace(jobId, std::move(jobCb), std::move(repeatInterval));
```
don't see why is this necessary? as a general thumb rule we should avoid acquiring mutex inside mutex.

veerajsk replied:
so the sequence that would fail I think is:
- Scheduler thread sees that's the `_jobQueue` is empty inside the `wait()` predicate i.e. `pop()` returns an empty optional
- Another thread adds a job to `_jobQueue` & calls `notify_one()`
- Scheduler thread misses the notify because it is still executing the `wait()` predicate & ends up waiting on the condition variable

Regarding the double lock, I am currently checking if we can get rid of the internal mutex in the queue.

veerajsk replied:
I got rid of the job queue's mutex

veerajsk committed:
Scheduler's job queue uses condition var's mutex

Remove the internal job queue mutex & use the scheduler's condition
variable mutex since this mutex has to be acquired in any case
while adding to the queue.

veerajsk commented:
> > > The job queue is not under the lock, are there performance concerns?
> > 
> > 
> > The queue itself is multi thread safe i.e. it has its own internal lock that it uses to guard access
> 
> Right, i mean with one more lock it's slower, would it affect someone?

I looked a bit more into the code & got rid of the internal mutex in the job queue, since we need to acquire the condition variable's mutex in any case while adding to the queue. So I used it for protecting access to the queue everywhere.

veerajsk committed:
Remove unused header in job queue

xffox commented on alica_engine/include/engine/scheduler/Scheduler.h:
```diff
@@ -89,8 +96,8 @@ class Scheduler
         while (_running) {
-            auto job = _jobQueue.pop();
-            if (!job) {
+            std::optional<Job> job;
```
Can you also add `optional` include for this?
